### PR TITLE
[FLINK-23848][connector/pulsar] Fix the consumer not found error in test.

### PR DESCRIPTION
**Pull request by syhily**
## What is the purpose of the change

Pulsar would disconnect the existed consumer when we perform a seek operation on a topic partition. If we have multiple consumers with the same subscription, we may meet a consumer not found error. That means Pulsar doesn't support seek on a partition with the same subscription concurrently. Move the seek operation to enumerator would make sure all the partitions would be executed one by one and only once. Thus fixed the test issue.

## Brief change log

  - Drop `StartCursor` in `PulsarPartitionSplit`.
  - Move `Consumer.seek` to source enumerator.

## Verifying this change

This change is already covered by existing tests, such as *PulsarSourceITCase*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (don't know)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/PulsarSourceBuilder.java [hunk 1]
```diff
@@ -473,12 +473,13 @@
...
         checkState(isSerializable(stopCursor), "StopCursor isn't serializable");
+        checkState(isSerializable(rangeGenerator), "RangeGenerator isn't serializable");
 
         // Make the configuration unmodifiable.
-        UnmodifiableConfiguration config = new UnmodifiableConfiguration(configuration);
+        UnmodifiableConfiguration immutableConfig = new UnmodifiableConfiguration(configuration);
 
         return new PulsarSource<>(
-                config,
+                immutableConfig,
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/PulsarSourceBuilder.java:482]: unnecessary changes

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java [hunk 2]
```diff
@@ -148,8 +162,44 @@ public void close() {
...
+    }
+
+    private void seekStartPosition(Set<TopicPartition> partitions) {
+        ConsumerBuilder<byte[]> consumerBuilder =
+                createConsumerBuilder(pulsarClient, Schema.BYTES, configuration);
+        Set<String> seekedPartitions = new HashSet<>();
+
+        for (TopicPartition partition : partitions) {
+            String topicName = partition.getFullTopicName();
+            if (!seekedPartitions.contains(topicName)) {
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java:182]:
```suggestion
            if (seekedPartitions.add(topicName)) {
```

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java [hunk 3]
```diff
@@ -148,8 +162,44 @@ public void close() {
...
+                        // WARN_ON_MISMATCH would just print this warning message.
+                        // No need to print the stacktrace.
+                        LOG.warn(
+                                "Failed to set initial consuming position for partition {}",
+                                partition,
+                                e);
+                    }
+                }
+
+                seekedPartitions.add(topicName);
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java:200]:
```suggestion
```

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java [hunk 4]
```diff
@@ -148,8 +162,44 @@ public void close() {
...
+        return partitions;
+    }
+
+    private void seekStartPosition(Set<TopicPartition> partitions) {
+        ConsumerBuilder<byte[]> consumerBuilder =
+                createConsumerBuilder(pulsarClient, Schema.BYTES, configuration);
+        Set<String> seekedPartitions = new HashSet<>();
+
+        for (TopicPartition partition : partitions) {
+            String topicName = partition.getFullTopicName();
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java:181]: When do we actually have duplicate topicNames? Is it for sticky keys?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java [hunk 5]
```diff
@@ -148,8 +162,44 @@ public void close() {
...
+                } catch (PulsarClientException e) {
+                    if (sourceConfiguration.getVerifyInitialOffsets() == FAIL_ON_MISMATCH) {
+                        throw new IllegalArgumentException(e);
+                    } else {
+                        // WARN_ON_MISMATCH would just print this warning message.
+                        // No need to print the stacktrace.
+                        LOG.warn(
+                                "Failed to set initial consuming position for partition {}",
+                                partition,
+                                e);
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java:196]: Get rid of `e` parameter? Or do we need the stacktrace?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReader.java [hunk 6]
```diff
@@ -75,7 +78,30 @@ protected void finishedPollMessage(Message<byte[]> message) {
...
+        MessageId latestConsumedId = split.getLatestConsumedId();
+
+        // Reset the start position for ordered pulsar consumer.
+        if (latestConsumedId != null) {
+            StartCursor startCursor = StartCursor.fromMessageId(latestConsumedId, false);
+            TopicPartition partition = split.getPartition();
+
+            try {
+                startCursor.seekPosition(
+                        partition.getTopic(), partition.getPartitionId(), consumer);
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReader.java:90]: Why do we seek here? I thought the purpose was to only seek in enumerator?

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarUnorderedPartitionSplitReader.java [hunk 7]
```diff
@@ -137,16 +137,18 @@ protected void startConsumer(PulsarPartitionSplit split, Consumer<byte[]> consum
...
     }
 
     public PulsarPartitionSplitState snapshotState(long checkpointId) {
-        TxnID txnID = PulsarTransactionUtils.getId(uncommittedTransaction);
-        this.uncommittedTransaction = newTransaction();
         PulsarPartitionSplitState state = new PulsarPartitionSplitState(registeredSplit);
-        state.setUncommittedTransactionId(txnID);
+
+        // Avoiding NP problem when Pulsar don't get the message before Flink checkpoint.
+        if (uncommittedTransaction != null) {
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarUnorderedPartitionSplitReader.java:147]: Please add `@Nullable` to `uncommittedTransaction` field declaration that helps IDE to warn us for such issues.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/split/PulsarPartitionSplitTest.java [hunk 8]
```diff
@@ -1,54 +0,0 @@
...
-import static org.apache.flink.connector.pulsar.source.enumerator.cursor.StartCursor.defaultStartCursor;
-import static org.apache.flink.connector.pulsar.source.enumerator.cursor.StopCursor.defaultStopCursor;
-import static org.apache.flink.connector.pulsar.source.enumerator.topic.TopicRange.createFullRange;
-import static org.junit.jupiter.api.Assertions.assertEquals;
-
-/** Unit tests for {@link PulsarPartitionSplit}. */
-class PulsarPartitionSplitTest {
-
-    @Test
-    void partitionSplitWouldHaveANewStartCursorWhenCheckpointing() {
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/split/PulsarPartitionSplitTest.java:38]: Why did you remove this test?

**Review by AHeise — COMMENTED** [review]
Thank you for the fix. I added a few remarks but it looks good overall.

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/split/PulsarPartitionSplitTest.java (hunk 8 above)
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/split/PulsarPartitionSplitTest.java:38]: This test was originally used to check if the StartCursor was updated when the PulsarPartitionSplit was created, but now the StartCursor is no longer used in the Split, so the corresponding test is no longer necessary.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReader.java (hunk 6 above)
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReader.java:90]: It's used only for `Exclusive` or `Failover` subscriptions. Only one consumer would be used for seeking the start position.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java [hunk 9]
```diff
@@ -148,8 +162,42 @@ public void close() {
...
      */
     private Set<TopicPartition> getSubscribedTopicPartitions() {
-        return subscriber.getSubscribedTopicPartitions(
-                pulsarAdmin, rangeGenerator, context.currentParallelism());
+        int parallelism = context.currentParallelism();
+        Set<TopicPartition> partitions =
+                subscriber.getSubscribedTopicPartitions(pulsarAdmin, rangeGenerator, parallelism);
+
+        // Seek start position for given partitions.
+        seekStartPosition(partitions);
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java:170]: I think this should only be called in `assignPendingPartitionSplits`. Now if you have discover enabled, it would actually seek too often.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReader.java (hunk 6 above)
Comment (AHeise) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReader.java:90]: In case of restart, isn't `latestConsumedId` always `null`? I thought we would update the `startCursor` on checkpoint, such that it should be properly initialized in enumerator/reader right?

**Review by AHeise — COMMENTED** [review]
I think we have an issue with non-exclusive topic partitions as is:
- If we have a full restart, it make sense to reinitialize on enumerator.
- For partial restarts (regional failover), I don't see how we can plausibly initialize the reader:
- If we initialize on enumerator, we interfere with non-failed subtasks.
- If we initilalize on reader, we get the current issues.
- If we don't initialize, it's probably incorrect as well

I'm assuming that we need to have some messages between the readers and the enumerator:
- The enumerator is not seeking at all.
- Readers seek for exclusive subscriptions.
- For non-exclusive subscriptions, a reader signals to enumerator that the topic partition needs initialization and the reader does not consume from that subscription initially.
- Enumerator sends to all readers with the same topic partition some pause message (this is needed in case of regional failover).
- Enumerator passes some initialization token to one of the readers that all want to seek the same topic partition.
- That one reader seeks and informs enumerator.
- Enumerator unpauses all readers of that TopicPartition.
- Readers that have been paused need to ignore all message ids < pause message id (in case of regional failover)

The protocol can probably be simplified in a few steps but I'm convinced this is needed.

**Comment by syhily** [comment]
> I think we have an issue with non-exclusive topic partitions as-is:
> 
> * If we have a full restart, it makes sense to reinitialize on the enumerator.
> * For partial restarts (regional failover), I don't see how we can plausibly initialize the reader:
> * If we initialize on enumerator, we interfere with non-failed subtasks.
> * If we initialize on the reader, we get the current issues.
> * If we don't initialize, it's probably incorrect as well

The issue of non-exclusive topic partitions doesn't exist. Let me clarify it. Consuming with the `Shared` or `Key_Shared` (on non-exclusive topic partitions) subscription would share the same partition among different readers. We can't use cumulative acknowledgment in these two modes, which means we have to mark the messages as read one by one.

Record all the consumed messages and marked as read after successfully performing checkpoint doesn't sound like a good design. So we use [Transaction](https://pulsar.apache.org/docs/en/txn-how/#3-acknowledge-messages-with-a-transaction) to solve this problem.

When we met partial restarts, the uncommitted Transaction would be abort, we would redeliver all the unacknowledged messages (the messages which don't mark read) by using `consumer.redeliverUnacknowledgedMessages()`.

**Comment by syhily** [comment]
> I'm assuming that we need to have some messages between the readers and the enumerator:
> 
> * The enumerator is not seeking at all.
> * Readers seek for exclusive subscriptions.
> * For non-exclusive subscriptions, a reader signals to enumerator that the topic partition needs initialization and the reader does not consume from that subscription initially.
> * Enumerator sends to all readers with the same topic partition some pause message (this is needed in case of regional failover).
> * Enumerator passes some initialization token to one of the readers that all want to seek the same topic partition.
> * That one reader seeks and informs enumerator.
> * Enumerator unpauses all readers of that TopicPartition.
> * Readers that have been paused need to ignore all message ids < pause message id (in case of regional failover)

This design is awesome. Move all the seek action in one place is better than the current solution. I have read the code. I think we need to implement the `SourceEvent`. But I may need some time to learn how to use it.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReader.java (hunk 6 above)
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/reader/split/PulsarOrderedPartitionSplitReader.java:90]: I have changed the serialization logic, the `latestConsumedId` shouldn't be `null` here when we restart.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java [hunk 10]
```diff
@@ -175,11 +175,11 @@ public void close() {
     private void seekStartPosition(Set<TopicPartition> partitions) {
         ConsumerBuilder<byte[]> consumerBuilder =
                 createConsumerBuilder(pulsarClient, Schema.BYTES, configuration);
-        Set<String> seekedPartitions = new HashSet<>();
+        Set<String> seekedTopics = new HashSet<>();
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java:178]: Since this is a local variable, you still end up seeking the topics on each discovery.
It might be better to switch the call of this function to `assign...` as this is only called on the newly discovered partitions.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java [hunk 11]
```diff
@@ -175,11 +175,11 @@ public void close() {
...
 
         for (TopicPartition partition : partitions) {
             String topicName = partition.getFullTopicName();
-            if (seekedPartitions.add(topicName)) {
+            if (!assignmentState.containsTopic(topicName) && seekedTopics.add(topicName)) {
```
Comment (AHeise) [inline flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java:182]: We could avoid this check by using only hooking into `assign` which only gets the added partitions. Did you try that out? (It might not work out, but it feels the better place).

Comment (AHeise) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java:182]: Clarified offline that this would not be as straight-forward and will be eventually superseded by the protocol sketched above.

**Review by AHeise — APPROVED** [review]
LGTM. Thank you very much!
